routers/repo/http.go

```diff
@@ -68,189 +52,188 @@ func HTTP(ctx *context.Context) {
 	if err != nil {
 		if models.IsErrUserNotExist(err) {
 			ctx.Handle(http.StatusNotFound, "GetUserByName", nil)
-		} else {
-			ctx.Handle(http.StatusInternalServerError, "GetUserByName", err)
+			return
 		}
+		ctx.Handle(http.StatusInternalServerError, "GetUserByName", err)
 		return
 	}
 
 	repo, err := models.GetRepositoryByName(repoUser.ID, reponame)
 	if err != nil {
 		if models.IsErrRepoNotExist(err) {
 			ctx.Handle(http.StatusNotFound, "GetRepositoryByName", nil)
-		} else {
-			ctx.Handle(http.StatusInternalServerError, "GetRepositoryByName", err)
+			return
 		}
+		ctx.Handle(http.StatusInternalServerError, "GetRepositoryByName", err)
 		return
 	}
 
-	// Only public pull don't need auth.
-	isPublicPull := !repo.IsPrivate && isPull
-	var (
-		askAuth      = !isPublicPull || setting.Service.RequireSignInView
-		authUser     *models.User
-		authUsername string
-		authPasswd   string
-		environ      []string
-	)
+	var environ []string
 
-	// check access
-	if askAuth {
+	// Check access
+	// Only public pull don't need auth.
+	if repo.IsPrivate || !isPull || setting.Service.RequireSignInView {
+		var reverseProxyAuth bool
+		var authHeader = ctx.Req.Header.Get("Authorization")
 		if setting.Service.EnableReverseProxyAuth {
-			authUsername = ctx.Req.Header.Get(setting.ReverseProxyAuthUser)
-			if len(authUsername) == 0 {
-				ctx.HandleText(401, "reverse proxy login error. authUsername empty")
-				return
-			}
-			authUser, err = models.GetUserByName(authUsername)
-			if err != nil {
-				ctx.HandleText(401, "reverse proxy login error, got error while running GetUserByName")
-				return
-			}
-		} else {
-			authHead := ctx.Req.Header.Get("Authorization")
-			if len(authHead) == 0 {
-				ctx.Resp.Header().Set("WWW-Authenticate", "Basic realm=\".\"")
-				ctx.Error(http.StatusUnauthorized)
+			reverseProxyAuth = true
+			authHeader = ctx.Req.Header.Get(setting.ReverseProxyAuthUser)
+		}
+		authUser, promptForAuth, httpStatus, err := checkAccess(repo, unitType, isPull, authHeader, reverseProxyAuth)
+		if err != nil {
+			if httpStatus == http.StatusInternalServerError {
+				ctx.Handle(httpStatus, "checkAccess", err)
 				return
 			}
+			ctx.HandleText(httpStatus, err.Error())
```

Comment:
You shouldn't return the real error information to the UI.

Reply:
In this case it's error that needs to be returned

Reply:
@lunny Error now contains text that was returned before, it has same behaviour from outside.